### PR TITLE
feat(tribeUsersWithMaxFollowers): #83 cloud function to get tribe users with Max followers count

### PR DESCRIPTION
**Pull request by VenkatVenkatesh**
- Implemented cloud functions to get tribe users with Maximum followers count.

- Used the cloud URL to obtain the response and display in the home page.
screenshot_2019-05-20_at_11 47 49_pm

### CODE REVIEW

File: functions/src/index.ts [hunk 1]
```diff
@@ -2,14 +2,22 @@ import * as functions from 'firebase-functions';
 import * as admin from 'firebase-admin';
 import { accountType } from './services/user/getAccountType';
 import { User } from './models/user.model';
+// import serviceAccount from './dashboard-dev.json'; // Replace the firebase json file here
```
Comment (eddiejaoude) [inline functions/src/index.ts:5]: Is this comment still needed?

File: functions/src/index.ts [hunk 2]
```diff
@@ -2,14 +2,22 @@ import * as functions from 'firebase-functions';
...
+import { fetchUser } from './services/user/fetchUser';
+import { UserSocial } from './models/userSocial.model';
 
-admin.initializeApp();
+admin.initializeApp({
+  // credential: admin.credential.cert((serviceAccount) as admin.ServiceAccount),
```
Comment (eddiejaoude) [inline functions/src/index.ts:10]: Is this comment still needed?

File: functions/src/index.ts [hunk 3]
```diff
@@ -2,14 +2,22 @@ import * as functions from 'firebase-functions';
...
+  databaseURL: "https://tribedashboard-dev.firebaseio.com"
+}
+);
 
 // Start writing Firebase Functions
 // https://firebase.google.com/docs/functions/typescript
 
 export const cron = functions.https.onRequest(async (request, response) => {
   try {
+    console.log('Entered into firbease functions');
```
Comment (eddiejaoude) [inline functions/src/index.ts]: Is this console still needed?

File: functions/src/services/user/fetchUser.ts [hunk 4]
```diff
@@ -64,7 +65,26 @@ const fetchTwitterRecord = (userName: string, document: User, provider: string)
   }
 }
 
+const calculateFollowersCount = (userSocial: UserSocial) => {
+  let githubCount = 0;
+  let twitterCount = 0;
+  let youtubeCount = 0;
+
+  if (userSocial['github']) {
```
Comment (eddiejaoude) [inline functions/src/services/user/fetchUser.ts]: Would this be better as `userSocial.github`?

File: functions/tsconfig.json [hunk 5]
```diff
@@ -6,6 +6,9 @@
     "outDir": "lib",
     "sourceMap": true,
     "strict": true,
+    "allowSyntheticDefaultImports": true,
+    "esModuleInterop": true,
+    "resolveJsonModule": true,
```
Comment (eddiejaoude) [inline functions/tsconfig.json:11]: Are these options still needed or just part of debugging?

File: web/src/app/components/home/home.component.html [hunk 6]
```diff
@@ -1,6 +1,54 @@
...
-    <button mat-flat-button class="mat-primary" (click)="navigateToLogin()">
-    OnBoard
-  </button>
+<div class="homepage-container" fxLayout="column" fxLayoutGap="10vh">
+  <div class="get-started-container">
+    <div>Your social tribe on a single dashboard</div>
+    <div class="primary-text">Connect your socials in one place</div>
+    <button class="get-started-button" mat-button cdkFocusInitial
+      (click)="navigateToLogin()">Get Started</button><br />
+    <img class="pattern-right" src="../../../assets/images/OvalPattern.svg" />
```
Comment (eddiejaoude) [inline web/src/app/components/home/home.component.html]: `assets` should be available publicly, these can be accessed a few ways, for example `assets:github` but not *relative*

File: web/src/app/components/home/home.component.html [hunk 7]
```diff
@@ -1,6 +1,54 @@
...
+    <mat-card *ngFor="let user of users" fxFlex="20">
+      <mat-card>
+        <div [ngSwitch]="getProvider(user)" class="avatar">
+          <div *ngSwitchCase="'github'">
+            <img [src]="user.github.avatarUrl" />
+            <div>User name</div>
+          </div>
+          <div *ngSwitchCase="'twitter'">
+            <img [src]="user.twitter.avatarUrl" />
+            <div>User name</div> <!-- Hard coded text will remove it in the next PR-->
```
Comment (eddiejaoude) [inline web/src/app/components/home/home.component.html]: Can you add a `@TODO:` in the comment so it doesnt get lost

File: web/src/app/components/home/home.component.scss [hunk 8]
```diff
@@ -0,0 +1,102 @@
+@import "../../../styles/variables.scss";
+
+.homepage-container {
+  .mat-card {
+    box-shadow: unset !important;
```
Comment (eddiejaoude) [inline web/src/app/components/home/home.component.scss]: I haven't seen the use of `!important` in a while, I think there might be a better way, where does the `box-shadow` come from?

File: web/src/app/components/home/home.component.ts [hunk 9]
```diff
@@ -1,18 +1,30 @@
...
   constructor(
-    private router: Router
-  ) { }
+    private router: Router,
+    private userService: UserService
+  ) {
+    this.userService.getUsersWithMaxFollowers()
+      .subscribe((value) => {
+        this.users = value;
+      });
```
Comment (eddiejaoude) [inline web/src/app/components/home/home.component.ts]: Doesn't need to be multi line function, can be...

```typescript
.subscribe((users: Users) => this.users = users);
```

*Note: the variable name and type 🤓*

File: web/src/app/services/user.service.ts [hunk 10]
```diff
@@ -197,4 +199,7 @@ export class UserService {
       .subscribe(() => this.router.navigate(['/dashboard'])
       );
   }
+  getUsersWithMaxFollowers() {
+    return this.http.get(PROVIDERS.TRIBE_COUNT_API);
```
Comment (eddiejaoude) [inline web/src/app/services/user.service.ts:203]: How is the `http` request protected?

File: functions/src/index.ts (hunk 1 above)
Comment (VenkatVenkatesh) [reply functions/src/index.ts:5]: @eddiejaoude I have added this comment for your reference.

File: functions/src/index.ts (hunk 3 above)
Comment (VenkatVenkatesh) [reply functions/src/index.ts]: Will remove it @eddiejaoude

File: functions/tsconfig.json (hunk 5 above)
Comment (VenkatVenkatesh) [reply functions/tsconfig.json:11]: As the index.ts file was throwing error for adding json file. I added these options. They are needed.

File: web/src/app/services/user.service.ts (hunk 10 above)
Comment (VenkatVenkatesh) [reply web/src/app/services/user.service.ts:203]: Planning to add auth headers. Will do it in the next PR.

**Comment by VenkatVenkatesh** [comment]
@eddiejaoude I addressed your question. Please let me know if i can resolve them if it sounds good to you.

File: functions/src/index.ts (hunk 1 above)
Comment (eddiejaoude) [reply functions/src/index.ts:5]: Ok, thank you, can you please add a `@TODO` so I remember to look at it

File: functions/tsconfig.json (hunk 5 above)
Comment (eddiejaoude) [reply functions/tsconfig.json:11]: ok 👍

File: web/src/app/components/home/home.component.ts [hunk 11]
```diff
@@ -1,18 +1,29 @@
...
+      .subscribe((tribeUsers: UserSocial[]) => this.users = tribeUsers);
+  }
 
   navigateToLogin(): void {
     this.router.navigate(['/login']);
+    this.userService.getUsersWithMaxFollowers();
+  }
+
+  getProvider(user): string {
+    return Object.keys(user)[0] === 'id' ? Object.keys(user)[1] : Object.keys(user)[0]; // Will remove the dummy field
```
Comment (eddiejaoude) [inline web/src/app/components/home/home.component.ts]: Is this comment still needed?

File: web/src/app/services/user.service.ts (hunk 10 above)
Comment (eddiejaoude) [reply web/src/app/services/user.service.ts:203]: Ok 👍 , might be easier to have it as a callable function?

**Comment by VenkatVenkatesh** [comment]
@eddiejaoude I have addressed the review comments. Please review and merge.

**Review by eddiejaoude — APPROVED** [review]

**Pull request merged**